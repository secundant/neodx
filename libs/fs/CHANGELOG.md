# @neodx/fs

## 0.0.8

### Patch Changes

- Updated dependencies [[`bc94f93`](https://github.com/secundant/neodx/commit/bc94f933baecc6dad4702359ebde5bc64b81141e), [`bc94f93`](https://github.com/secundant/neodx/commit/bc94f933baecc6dad4702359ebde5bc64b81141e)]:
  - @neodx/std@0.1.4

## 0.0.7

### Patch Changes

- Updated dependencies [[`b5dc26e`](https://github.com/secundant/neodx/commit/b5dc26e0c47bab3f01863a15cf5db360e39efdcf)]:
  - @neodx/std@0.1.3

## 0.0.6

### Patch Changes

- Updated dependencies [[`2c1a1a6`](https://github.com/secundant/neodx/commit/2c1a1a6e1e2980f2fdd26260790707db7352bce8), [`2c1a1a6`](https://github.com/secundant/neodx/commit/2c1a1a6e1e2980f2fdd26260790707db7352bce8), [`e2f2ece`](https://github.com/secundant/neodx/commit/e2f2eceec7aaa5d4ddbc5f156a018cf3822d529a), [`e2f2ece`](https://github.com/secundant/neodx/commit/e2f2eceec7aaa5d4ddbc5f156a018cf3822d529a)]:
  - @neodx/std@0.1.2

## 0.0.5

### Patch Changes

- Updated dependencies [[`3d2caaa`](https://github.com/secundant/neodx/commit/3d2caaa792bcf392765c08d11c6a82c3a19295e4)]:
  - @neodx/std@0.1.1

## 0.0.4

### Patch Changes

- [#47](https://github.com/secundant/neodx/pull/47) [`248f15a`](https://github.com/secundant/neodx/commit/248f15ab83719f4fecc19c6882442c8815d3bfba) Thanks [@secundant](https://github.com/secundant)! - Added JSON utilities, though the API is not yet stable

- Updated dependencies [[`ea3119d`](https://github.com/secundant/neodx/commit/ea3119d23cd6107b1a1c71caf69877511536a975), [`7676903`](https://github.com/secundant/neodx/commit/76769036464164b3db2b9ff13a63b72e719430e6), [`8264689`](https://github.com/secundant/neodx/commit/826468971ee171e5e2a0a28c55e0a2e9411f12a3), [`248f15a`](https://github.com/secundant/neodx/commit/248f15ab83719f4fecc19c6882442c8815d3bfba)]:
  - @neodx/std@0.1.0

## 0.0.3

### Patch Changes

- [#33](https://github.com/secundant/neodx/pull/33) [`246f4f2`](https://github.com/secundant/neodx/commit/246f4f292a005be440d78e7528cc40aefa5c6ad8) Thanks [@secundant](https://github.com/secundant)! - Fix incorrect versions in dependencies (workspace:\*)

- Updated dependencies [[`246f4f2`](https://github.com/secundant/neodx/commit/246f4f292a005be440d78e7528cc40aefa5c6ad8)]:
  - @neodx/std@0.0.3

## 0.0.2

### Patch Changes

- [#25](https://github.com/secundant/neodx/pull/25) [`db1e519`](https://github.com/secundant/neodx/commit/db1e5193c4c5af6e0583a5e2f2e0a2ff161208d6) Thanks [@secundant](https://github.com/secundant)! - Update builder

- Updated dependencies [[`db1e519`](https://github.com/secundant/neodx/commit/db1e5193c4c5af6e0583a5e2f2e0a2ff161208d6)]:
  - @neodx/std@0.0.2
