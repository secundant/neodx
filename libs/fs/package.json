{
  "name": "@neodx/fs",
  "version": "0.0.2",
  "description": "File system helpers",
  "type": "module",
  "main": "./dist/index.cjs",
  "types": "./dist/index.d.ts",
  "module": "./dist/index.mjs",
  "source": "src/index.ts",
  "sideEffects": false,
  "files": [
    "dist",
    "README.md",
    "package.json"
  ],
  "scripts": {
    "typecheck": "tsc --noEmit",
    "clean": "yarn dlx rimraf dist",
    "build": "autobuild",
    "test": "vitest run",
    "lint": "eslint src"
  },
  "devDependencies": {
    "@neodx/autobuild": "workspace:*",
    "@types/node": "18.11.18",
    "@types/tmp": "^0.2.3",
    "eslint": "8.31.0",
    "eslint-kit": "6.10.0",
    "tmp": "0.2.1",
    "typescript": "4.9.4"
  },
  "dependencies": {
    "@neodx/std": "workspace:*",
    "tiny-glob": "0.2.9"
  },
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts"
    }
  }
}
