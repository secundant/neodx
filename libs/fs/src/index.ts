export { assertDir, assertFile, exists, isDirectory, isFile, isValidStats } from './checks';
export { ensureDir, ensureFile } from './ensure';
export type { ParseJsonParams, SerializeJsonParams } from './json';
export { parseJson, serializeJson } from './json';
export { deepReadDir } from './read';
export type { ScanParams } from './scan';
export { scan } from './scan';

// re-exported from fs/promises

export {
  access,
  appendFile,
  chmod,
  chown,
  copyFile,
  cp,
  lstat,
  mkdir,
  mkdtemp,
  opendir,
  readdir,
  readFile,
  rename,
  rm,
  stat,
  unlink,
  watch,
  writeFile
} from 'node:fs/promises';
