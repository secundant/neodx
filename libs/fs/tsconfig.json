{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "allowSyntheticDefaultImports": true
  }
}
