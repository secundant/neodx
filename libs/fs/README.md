# @neodx/fs

File system enhancements and common stuff.

API`s overview:

- `scan(path, ['*.js', '*.module.ts', '!*.test.{ts,js}'])` - glob reader with multiple inputs support
- `ensureDir(path)`, `ensureFile(path)` - Safe file/dir creation with all required ancestors
- `exists(path)`, `isFile(path)`, `isDirectory(path)` - Safe path checks
- `deepReadDir(path)` - Deep version of `readdir`
- `parseJson(input)`, `serializeJson(input)` - Safe JSON parser and serializer with [JSONC](https://www.npmjs.com/package/jsonc-parser) support

## API

### scan

Glob-based (via [tiny-glob](https://www.npmjs.com/package/tiny-glob)) multiple patterns scanner with exclusion support

```typescript
import { scan } from '@neodx/fs';

await scan(process.cwd(), ['*.js', '!*.config.js']);
await scan(process.cwd(), '**/*.ts', '**/*.js');
```

### `ensureFile(path)` and `ensureDir(path)`

Recursively creates missed file or dir with all required ancestors if one of them is not exists.

Automatically avoid duplicated calls:

```typescript
import { ensureFile, ensureDir } from '@neodx/fs';

// Everything works as expected
await Promise.all([
  ensureDir('foo/baz'),
  ensureFile('foo/bar/2.ts'),
  ensureDir('foo/bar'),
  ensureFile('foo/bar/1.ts'),
  ensureDir('foo')
]);
```

### `isFile`, `isDirectory`, `exists`

The following APIs are useful for safe paths checking.

- `exists(path)` - Returns `true` if the path exists
- `isFile(path)` - Returns `true` if the path exists, and it's a file
- `isDirectory(path)` - Returns `true` if the path exists, and it's a directory

### `deepReadDir(path, { absolute = false })`

Returns flat list with all children's paths. Paths are relative by default.

```typescript
import { deepReadDir, isFile } from '@neodx/fs';

const files = await deepReadDir(myPath);

for (const file of files) {
  if (await isFile(file)) {
    await doSmth(file);
  }
}
```

### `parseJson` and `serializeJson`

```typescript
import { parseJson, serializeJson } from '@neodx/fs';
import { writeFile, readFile } from 'fs/promises';

const json = parseJson(await readFile('tsconfig.json', 'utf-8'));

await writeFile(
  'tsconfig.json',
  serializeJson({
    ...json,
    compilerOptions: {
      ...json.compilerOptions,
      target: 'esnext'
    }
  }),
  'utf-8'
);
```

---

Inspired by fs-extra and others.
