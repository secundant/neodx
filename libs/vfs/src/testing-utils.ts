import { writeFile } from 'fs/promises';
import { ensureFile, getHash } from '@neodx/fs';
import { concurrently, entries } from '@neodx/std';
import { resolve } from 'pathe';
import type { Options } from 'prettier';
import { dirSync } from 'tmp';
import { type CreateVfsParams, type VFS, createVfs } from './create-vfs';
import {
  type VirtualFsInitializer,
  flattenVirtualFsInitializer
} from './implementations/virtual-fs';
import type { ContentLike } from './types';

export interface TmpVfsParams extends Omit<CreateVfsParams, 'root'> {
  initialFiles?: VirtualFsInitializer;
}

export async function createTmpVfs({ initialFiles, ...options }: TmpVfsParams = {}) {
  const vfs = createVfs(dirSync().name, options);

  if (initialFiles) {
    await writeRealFs(vfs.root, flattenVirtualFsInitializer(initialFiles, vfs.root));
  }
  return vfs;
}

export async function getChangesHash(vfs: VFS) {
  const changes = await vfs.getChanges();

  return changes.map(({ name, content }) => [name, content && getHash(content)]);
}

// TODO Move to @neodx/fs
export async function writeRealFs(
  root: string,
  target: Map<string, ContentLike> | Record<string, ContentLike>
) {
  await concurrently(
    target instanceof Map ? target : entries(target),
    async ([file, content]) => {
      await ensureFile(resolve(root, file));
      await writeFile(resolve(root, file), content);
    },
    10
  );
}

export async function writeVfsPackageConfiguration(vfs: VFS, name = 'my-pkg') {
  await vfs.writeJson<Options>('.prettierrc', { singleQuote: true, printWidth: 80 });
  await vfs.writeJson('package.json', {
    name
  });
  await vfs.applyChanges();
  return vfs;
}
