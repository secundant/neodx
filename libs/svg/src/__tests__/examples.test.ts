import { deepReadDir } from '@neodx/fs';
import { describe, expect, test } from 'vitest';
import type { GenerateParams } from '../index';
import { generateExample, getExamplesNames } from './testing-utils';

describe('examples', () => {
  const examples = getExamplesNames();
  const optionsMap: Record<string, Partial<GenerateParams>> = {
    'groups-without-root': {
      group: true
    },
    'groups-with-root': {
      group: true,
      input: ['**/*.svg'],
      root: 'assets'
    },
    react: {
      group: true,
      input: ['**/*.svg'],
      root: 'assets'
    }
  };

  test.each(examples)(`"%s" example should replay same output`, async name => {
    const { tree } = await generateExample(name, false, optionsMap[name]);
    const originalChanges = await tree.getChanges();
    const serializedChanges = originalChanges.map(change => ({
      ...change,
      content: change.content?.toString('utf-8')
    }));

    expect(serializedChanges).toMatchSnapshot();
  });

  test.each(examples)(`"%s" example should generate files`, async name => {
    const { tree } = await generateExample(name, true, optionsMap[name]);

    expect(await deepReadDir(tree.root, { absolute: false })).toMatchSnapshot();
  });
});
