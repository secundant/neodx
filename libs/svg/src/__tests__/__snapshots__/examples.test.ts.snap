// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`examples > "colors" example should generate files 1`] = `
[
  "assets/custom.svg",
  "assets/fill.svg",
  "assets/mixed.svg",
  "assets/stroke.svg",
  "generated/sprite-info.ts",
  "generated/sprite.svg",
]
`;

exports[`examples > "colors" example should replay same output 1`] = `[]`;

exports[`examples > "groups-with-root" example should generate files 1`] = `
[
  "assets/common/close.svg",
  "assets/common/favourite.svg",
  "assets/format/align-left.svg",
  "assets/format/tag.svg",
  "generated/common.svg",
  "generated/format.svg",
  "generated/sprite-info.ts",
]
`;

exports[`examples > "groups-with-root" example should replay same output 1`] = `[]`;

exports[`examples > "groups-without-root" example should generate files 1`] = `
[
  "assets/common/close.svg",
  "assets/common/favourite.svg",
  "assets/format/align-left.svg",
  "assets/format/tag.svg",
  "generated/assets/common.svg",
  "generated/assets/format.svg",
  "generated/sprite-info.ts",
]
`;

exports[`examples > "groups-without-root" example should replay same output 1`] = `[]`;

exports[`examples > "react" example should generate files 1`] = `
[
  "assets/common/close.svg",
  "assets/common/favourite.svg",
  "assets/format/align-left.svg",
  "assets/format/tag.svg",
  "example.tsx",
  "generated/common.svg",
  "generated/format.svg",
  "generated/sprite-info.ts",
  "icon.tsx",
]
`;

exports[`examples > "react" example should replay same output 1`] = `[]`;

exports[`examples > "simple" example should generate files 1`] = `
[
  "assets/arrow-drop-down.svg",
  "assets/arrow-drop-up.svg",
  "generated/sprite-info.ts",
  "generated/sprite.svg",
]
`;

exports[`examples > "simple" example should replay same output 1`] = `[]`;
